packages/peregrine/src/Toasts/useToastContext.js

```diff
@@ -0,0 +1,127 @@
+import React, { createContext, useContext, useReducer } from 'react';
+/**
+ * The state of the toast store.
+ * @typedef {Object} ToastState
+ * @property {Object} toasts - unordered map of id:toast
+ */
+const initialState = {
+    toasts: {}
+};
+
+const reducer = (prevState = initialState, action = {}) => {
+    const { type, payload } = action;
+
+    if (!type) {
+        throw new TypeError('action.type is required');
+    }
+
+    if (!payload.id) {
+        throw new TypeError('action.payload.id is required');
+    }
+
+    switch (type) {
+        case 'add': {
+            // If we are adding a toast that already exists we need to clear the
+            // old removal timeout effectively resetting the delete timer.
+            if (prevState.toasts[payload.id]) {
+                window.clearTimeout(
+                    prevState.toasts[payload.id].removalTimeoutId
+                );
+            }
+
+            const duplicate = !!prevState.toasts[payload.id];
+
+            // For duplicate toasts, do not update the timestamp to maintain
+            // order of toast emission.
+            const timestamp = duplicate
+                ? prevState.toasts[payload.id].timestamp
+                : Date.now();
+
+            // Use a random key to trigger a recreation of this component if it
+            // is a duplicate so that we can re-trigger the blink animation.
+            const key = duplicate ? Math.random() : payload.id;
```

Comment:
Is using `Date.now()` to create the timestamp also bad form?

Another idea: 

1) Move `timestamp` creation to the action and use that as the ID for the toast store. 
2) In the reducer, if duplicate, reset the timestamp to the previous timestamp (otherwise leave it).
3) In the display, `toastContainer`, if we see `duplicate: true`, generate a random key.

This would still have a side-effect of modifying the incoming `timestamp` value in the event of a duplicate...

Reply:
> Is using Date.now() to create the timestamp also bad form?

Yes. A reducer should be idempotent: call it with the same arguments (`payload` and `type`) and it should return the same result every time. That won't happen if something in the reducer calls `Math.random()` or `Date.now()`, since those return a different value on every call.

The solution is usually to handle randomness in the action creator.